exporter/prometheusexporter/accumulator.go

```diff
@@ -63,104 +63,103 @@ func newAccumulator(logger *zap.Logger, metricExpiration time.Duration) accumula
 
 // Accumulate stores one datapoint per metric
 func (a *lastValueAccumulator) Accumulate(rm pdata.ResourceMetrics) (n int) {
+	now := time.Now()
 	ilms := rm.InstrumentationLibraryMetrics()
 
 	for i := 0; i < ilms.Len(); i++ {
 		ilm := ilms.At(i)
 
 		metrics := ilm.Metrics()
 		for j := 0; j < metrics.Len(); j++ {
-			n += a.addMetric(metrics.At(j), ilm.InstrumentationLibrary())
+			n += a.addMetric(metrics.At(j), ilm.InstrumentationLibrary(), now)
 		}
 	}
 
 	return
 }
 
-func (a *lastValueAccumulator) addMetric(metric pdata.Metric, il pdata.InstrumentationLibrary) int {
+func (a *lastValueAccumulator) addMetric(metric pdata.Metric, il pdata.InstrumentationLibrary, now time.Time) int {
 	a.logger.Debug(fmt.Sprintf("accumulating metric: %s", metric.Name()))
 
 	switch metric.DataType() {
 	case pdata.MetricDataTypeIntGauge:
-		return a.accumulateIntGauge(metric, il)
+		return a.accumulateIntGauge(metric, il, now)
 	case pdata.MetricDataTypeIntSum:
-		return a.accumulateIntSum(metric, il)
+		return a.accumulateIntSum(metric, il, now)
 	case pdata.MetricDataTypeDoubleGauge:
-		return a.accumulateDoubleGauge(metric, il)
+		return a.accumulateDoubleGauge(metric, il, now)
 	case pdata.MetricDataTypeDoubleSum:
-		return a.accumulateDoubleSum(metric, il)
+		return a.accumulateDoubleSum(metric, il, now)
 	case pdata.MetricDataTypeIntHistogram:
-		return a.accumulateIntHistogram(metric, il)
+		return a.accumulateIntHistogram(metric, il, now)
 	case pdata.MetricDataTypeHistogram:
-		return a.accumulateDoubleHistogram(metric, il)
+		return a.accumulateDoubleHistogram(metric, il, now)
 	}
 
 	return 0
 }
 
-func (a *lastValueAccumulator) accumulateIntGauge(metric pdata.Metric, il pdata.InstrumentationLibrary) (n int) {
+func (a *lastValueAccumulator) accumulateIntGauge(metric pdata.Metric, il pdata.InstrumentationLibrary, now time.Time) (n int) {
 	dps := metric.IntGauge().DataPoints()
 	for i := 0; i < dps.Len(); i++ {
 		ip := dps.At(i)
 
-		ts := ip.Timestamp().AsTime()
 		signature := timeseriesSignature(il.Name(), metric, ip.LabelsMap())
 
 		v, ok := a.registeredMetrics.Load(signature)
 		if !ok {
 			m := createMetric(metric)
 			m.IntGauge().DataPoints().Append(ip)
-			a.registeredMetrics.Store(signature, &accumulatedValue{value: m, instrumentationLibrary: il, stored: time.Now()})
+			a.registeredMetrics.Store(signature, &accumulatedValue{value: m, instrumentationLibrary: il, updated: now})
 			n++
 			continue
 		}
 		mv := v.(*accumulatedValue)
 
-		if ts.Before(mv.value.IntGauge().DataPoints().At(0).Timestamp().AsTime()) {
+		if ip.Timestamp().AsTime().Before(mv.value.IntGauge().DataPoints().At(0).Timestamp().AsTime()) {
```

Comment:
Unrelated to your change: `(mv.value.IntGauge().DataPoints().At(0).Timestamp().AsTime()` doesn't look very safe. I wonder if any term here can be nil and cause a panic.